File: Graphs/Paths & Cycles/Topological Sort/210. Course Schedule II.cpp
/*
 ____  _     _           _     _       __  __       _ _           _
|  _ \(_)___| |__   __ _| |__ | |__   |  \/  | __ _| | |__   ___ | |_ _ __ __ _
| |_) | / __| '_ \ / _` | '_ \| '_ \  | |\/| |/ _` | | '_ \ / _ \| __| '__/ _` |
|  _ <| \__ \ | | | (_| | |_) | | | | | |  | | (_| | | | | | (_) | |_| | | (_| |
|_| \_\_|___/_| |_|\__,_|_.__/|_| |_| |_|  |_|\__,_|_|_| |_|\___/ \__|_|  \__,_|

*/

#include <bits/stdc++.h>
using namespace std;

enum State
{
  UNVISITED,
  PROCESSING,
  VISITED
};

class Solution
{
private:
  vector<int> result;
  bool acyclic(int &node, vector<list<int>> &adjList, vector<State> &visited)
  {
    if (visited[node] == VISITED)
      return true;

    visited[node] = PROCESSING;

    for (int neighbour : adjList[node])
    {
      if (visited[neighbour] == PROCESSING)
        return false;
      if (visited[neighbour] == UNVISITED && !acyclic(neighbour, adjList, visited))
        return false;
    }
    visited[node] = VISITED;
    result.push_back(node);
    return true;
  }

public:
  vector<int> findOrder(int numCourses, vector<vector<int>> &prerequisites)
  {
    vector<list<int>> adjList(numCourses);

    for (vector<int> &course : prerequisites)
    {
      int u = course[0], v = course[1];
      adjList[v].push_back(u);
    }

    vector<State> visited(numCourses, UNVISITED);
    for (int node = 0; node < numCourses; node++)
      if (visited[node] == false && acyclic(node, adjList, visited) == false)
        return {};

    reverse(result.begin(), result.end());
    return result;
  }
};

File: Graphs/Paths & Cycles/Topological Sort/210. Course Schedule II.go
package main

const UNVISTED, ENQUEUE, VISITED int = 0, 1, 2

func buildGraph(edgeList [][]int, n int) [][]int {
	graph := make([][]int, n)
	for _, edge := range edgeList {
		// edge[1] -> edge[0]
		u, v := edge[1], edge[0]
		graph[u] = append(graph[u], v)
	}
	return graph
}

/*
 1 -> 2 -> 3
      |    |
			v		 V
			5	-> 4

	4 -> 5 -> 3 -> 2 -> 1
*/

func acyclic(course int, graph [][]int, visit []int, schedule *[]int) bool {
	visit[course] = ENQUEUE

	for _, nextCourse := range graph[course] {
		if visit[nextCourse] == ENQUEUE {
			return false
		}
		if visit[nextCourse] == UNVISTED && !acyclic(nextCourse, graph, visit, schedule) {
			return false
		}
	}

	*schedule = append(*schedule, course)

	visit[course] = VISITED

	return true
}

// 1 2 3 4 5
func reverse(A []int) []int {
	n := len(A)
	for i, j := 0, n-1; i < j; i, j = i+1, j-1 {
		A[i], A[j] = A[j], A[i]
	}
	return A
}

func findOrder(numCourses int, prerequisites [][]int) []int {

	graph := buildGraph(prerequisites, numCourses)
	visit := make([]int, numCourses)

	var schedule []int

	for i := range visit {
		visit[i] = UNVISTED
	}

	for course := range graph {
		if visit[course] == UNVISTED {
			if acyclic(course, graph, visit, &schedule) == false {
				return []int{}
			}
		}
	}
	schedule = reverse(schedule)

	return schedule
}
